File: 0670. Maximum Swap/main.js
const maximumSwap = function (num) {
  const swap = (s, i, j) => {
    let arr = s.split("");
    const c = arr[i];
    arr[i] = arr[j];
    arr[j] = c;
    return arr.join("");
  };

  let numArr = new Array(10).fill(-1);
  let s = num.toString();
  for (let i = 0; i < s.length; i++) {
    // console.log(parseInt(s[i]));
    numArr[parseInt(s[i])] = i;
  }
  for (let i = 0; i < s.length; i++) {
    for (let j = 9; j >= 0; j--) {
      if (numArr[j] != -1 && numArr[j] > i && j > parseInt(s[i])) {
        s = swap(s, i, numArr[j]);
        return new Number(s);
      }
    }
  }
  return num;
};

const num = 9973;
console.log(maximumSwap(num));

File: 0670. Maximum Swap/main.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

/**
 * I: num = 2736
 * O: 7236
 * Explanation: Swap the number 2 and the number 7.
 *
 * I: num = 9973
 * O: 9973
 * Explanation: No swap.
 */

int maximumSwap(int num);

void swap(string& s, int i, int j);

int main(int argc, char const* argv[]) {
    int num = 2736;
    cout << maximumSwap(num) << endl;
    return 0;
}

int maximumSwap(int num) {
    vector<int> numArr(10, -1);
    string s = to_string(num);
    for (int i = 0; i < s.length(); i++) {
        numArr[s[i] - '0'] = i;
    }
    for (int i = 0; i < s.length(); i++) {
        for (int j = 9; j >= 0; j--) {
            if (numArr[j] != -1 && numArr[j] > i && j > s[i] - '0') {
                swap(s, i, numArr[j]);
                return stoi(s);
            }
        }
    }

    return num;
}

void swap(string& s, int i, int j) {
    // cout << i << j << endl;
    char c = s[i];
    s[i] = s[j];
    s[j] = c;
}
